enhancements/kube-apiserver/generic-delete-protection.md

```diff
@@ -0,0 +1,233 @@
+   If all .spec.criteria are satisfied, ALLOW the delete.
+   If not all .spec.criteria are satisfied, do something sane in spec and DENY the delete.
+3. As a special case, a CriticalService cannot be deleted until its .spec.provider is no longer present.
+
+```go
+package api
+
+type CriticalService struct{
+    Spec CriticalServiceSpec
+}
+
+type CriticalServiceSpec struct{
+    Provider CriticalServiceProvider
+    Criteria []CriticalServiceCriteria
+}
+
+type GroupResource struct{
+    Group string
+    Resource string
+}
+
+type CriticalServiceProvider struct{
+    // only allow deployments.apps to start
+    GroupResource
+    Namespace string
+    Name string
+}
+
+type CriticalServiceCriteriaType string
+var(
+    FinalizerType CriticalServiceCriteriaType = "Finalizer"
+    SpecificResourceType CriticalServiceCriteriaType = "SpecificResource"
+)
+
+type CriticalServiceCriteria struct{
+    Type CriticalServiceCriteriaType
+    Finalizer *FinalizerCriticalServiceCriteria
+    SpecificResource *SpecificResourceCriticalServiceCriteria
+}
+
+type FinalizerCriticalServiceCriteria struct{
+    GroupResource
+    FinalizerName string
+}
+
+type SpecificResourceCriticalServiceCriteria struct{
+    GroupResource
+    Namespace string
+    Name string
+}
+```
+
+### User Stories 
+
+#### Deployment provides finalizer processing for CRD
+```yaml
+kind: CriticalService
+spec:
+  provider:
+    group: apps
+    resource: deployments
+    namespace: finalizer-namespace
+    name: finalizer-deployment
+  criteria:
+  - type: Finalizer
+    finalizer:
+      group: my.crd.group
+      resource: coolresources
+      finalizerName: my.crd.group/super-important
+  - type: SpecificResource
+    specificResource:
+      group: apiextensions.k8s.io
+      resource: CustomResourceDefinition
+      name: coolresources.my.crd.group
+```
+
+
+This would allow three separate ManifestWorks to all be deleted at the same time and avoid conflicting with each other
+as the deletes are finalized on individual resources.
+```yaml
+kind: ManifestWork
+metadata:
+  name: finalizer
+spec:
+  criticalservices/for-finalizer-deployment
+  namespace/finalizer-namespace
+  deployment.apps/finalizer-deployment
+---
+kind: ManifestWork
+metadata:
+  name: crd
+spec:
+  crd/coolresources.my.crd.group
+---
+kind: ManifestWork
+metadata:
+  name: cr
+spec:
+  coolresources.my.crd.group/some-instance
+```
+
+It would also allow them to be combined into a single ManifestWork.
+```yaml
+kind: ManifestWork
+metadata:
+  name: all
+spec:
+  criticalservices/for-finalizer-deployment
+  namespace/finalizer-namespace
+  deployment.apps/finalizer-deployment
+  crd/coolresources.my.crd.group
+  coolresources.my.crd.group/some-instance
+```
+
+This construct also means that namespaces in a management cluster can be be deleted when managed clusters are removed
+because deletion can happen in any order. 
+
+
+When a bulk delete happens, the effective order will be
```

Comment:
To confirm, the bulk delete happens in several loops,  that in each  loop some resource deletion is `DENIED` until all resources are deleted.